### PR TITLE
LKJ prior distribution for correlation matrices.

### PR DESCRIPTION
**Pull request by kiudee**
## Introduction

This is my first draft of the LKJ (Lewandowski, Kurowicka and Joe) distribution.
Since I am very busy with my master’s thesis, I wanted to publish the initial code to get some feedback.

The [LKJ (Lewandowski, Kurowicka and Joe)](http://www.sciencedirect.com/science/article/pii/S0047259X09000876) is a prior distribution for correlation matrices.
For n=1 this corresponds to the uniform distribution over correlation matrices.
If n larger than 1, the LKJ prior approaches the identity matrix.
## What works

The algorithm shown here has been implemented in [emcee](http://dan.iel.fm/emcee/current/) and tested in a multivariate normal setting.
## Problems

The PyMC implementation currently does not run. This is due to a problem with the `all(...)` statements which has been already mentioned here: #420
Furthermore, these checks are not even necessary if only the upper triangular matrix would be sampled.
(That is also what I did for the current implementation in emcee)

Any suggestion on what would be a good solution is welcome.

### CODE REVIEW

**Comment by twiecki** [comment]
Looks great! Regarding the error of `all`, are you sure it's using `theano`'s `all`? Maybe they just haven't implemented the gradient for that. Maybe @nouiz can advise?

**Comment by kiudee** [comment]
I looked at the current theano master branch and it turns out that the `elemwise.All` object does not provide gradient information.
Since the element wise `and` can be rewritten as a product, I assume the gradient information of the `Prod` object could be used.
But as is already apparent when reading the description of the `Prod.grad` function, this will result in a gradient of 0 in all cases where more than one subcondition is not satisfied.

**Comment by twiecki** [comment]
@kiudee Thanks for looking into that. Above you say that the check should basically never fail as the matrix you create is triangular or did I misread that?

**Comment by kiudee** [comment]
Oh, I am sorry for the confusion. In this implementation I expect the sampler to provide the complete matrix and will return `-np.inf` if it is not a correct correlation matrix.

In my `emcee` implementation I do not have any correctness checks. There, the sampler provides a 1-dimensional array of size `n * (n - 1) / 2` corresponding to the upper triangular matrix.
The user then has to make sure to correctly construct the correlation matrix:

``` python
# corr_raw is the 1d array provided by the sampler
corr = np.eye(ndim)
corr[np.triu_indices(ndim, k=1)] = corr_raw
corr.T[np.triu_indices(ndim, k=1)] = corr_raw
```

The ‘ideal’ solution would be that the probability distribution indicates an `input_shape` to the sampler, while providing a callable function to produce the indicated `shape`.

**Comment by twiecki** [comment]
I see. Can you add a simple example to the examples folder that demonstrates its usage?

File: pymc/distributions/multivariate.py [hunk 1]
```diff
@@ -177,3 +177,63 @@ def logp(self, X):
              2) - n * log(IVI) - 2 * multigammaln(p, n / 2)) / 2,
 
              n > (p - 1))
+
+
+class LKJCorr(Continuous):
+    """
+    The LKJ (Lewandowski, Kurowicka and Joe) is a prior distribution for
+    correlation matrices. For n=1 this corresponds to the uniform distribution
+    over correlation matrices. When n>1 the LKJ prior approaches the identity
```
Comment (twiecki) [inline pymc/distributions/multivariate.py]: For n->oo instead?

**Comment by kiudee** [comment]
As soon as I have larger break, I will write a small example showing how each of the described implementations would have to be used in a multivariate model.

**Comment by nouiz** [comment]
For Theano tensor.all(), as it was said, it do not implement gradient. But
as it return an int8, I'm not sure it make sence to implement it. Normally,
we implement gradient only on real number. @lamblin or @carriepl, what do
you think?

For the grad of Prod, it was a deliberate choice to make it return 0 if
more then 1 reduced group are 0. Why this is bad for you? There is the
ProdWithoutZeros, maybe this could help you. When it reduce 2 numbers, if
one of them is 0, is return the other number instead of multipling both
number.

**Comment by lamblin** [comment]
@nouiz Yes, it makes sense to implement the gradient in that case. Since the output is an int, the gradient wrt the input should be an explicit 0, in a floating-point type (floatX for the inputs that are integers).

**Comment by kiudee** [comment]
I added a small example which uses the LKJ prior as it is currently implemented.
Note, that this will not run, because we first need the gradient for `tensor.all()`.

**Comment by twiecki** [comment]
@nouiz Is there a chance that the gradient for `tensor.all()` would be added?

**Comment by nouiz** [comment]
I'm at GTC this week. Maybe @lamblin or @carriepl can comment? How hard it
will be? Can you explain what need to be done?

**Comment by lamblin** [comment]
That's trivial to add, I'll try to add it with a unit test today or tomorrow.

**Comment by lamblin** [comment]
Submitted in https://github.com/Theano/Theano/pull/2636

**Comment by kiudee** [comment]
@lamblin Thanks, that was quick!

**Comment by kiudee** [comment]
Progress update: I’ve verified that the LKJ prior implemented in Theano returns the correct values (matching my emcee implementation).
Unfortunately, pymc3 does only sample the identity matrix, which indicates to me that the sampler never finds any other matrix it accepts (too many constraints).

**Comment by kiudee** [comment]
Good news! I’ve re-implemented the LKJ prior as a prior for the upper triangular matrix.
This has the advantage that we do not need the symmetry constraints.

The only disadvantage is that the user has to manage the upper triangular matrix:

``` python
p = 4
n_elem = p * (p-1) / 2
tri_index = np.zeros([p,p], dtype=int)
tri_index[np.triu_indices(p, k=1)] = np.arange(n_elem)
tri_index[np.triu_indices(p, k=1)[::-1]] = np.arange(n_elem)
with Model() as model:
    mu = Normal('mu', mu=0, tau=1**-2, shape=p)
    sigma = Uniform('sigma', shape=p)
    corr_triangle = LKJCorrTriangle('corr', n=1, p=p)
    corr = corr_triangle[tri_index]
    corr = t.fill_diagonal(corr, 1)
    cov = Deterministic('cov', t.diag(sigma).dot(corr.dot(t.diag(sigma))))
    like = MvNormal('like', mu=mu, tau=inv(cov), observed=dataset)
```

Here is a traceplot of a simple multivariate normal model:
Trace of working correlation matrix

I’ll clean up the code later and commit it.

PS: By the way, the MvNormal distribution works as expected.

**Comment by twiecki** [comment]
@kiudee Awesome! Those posteriors look perfect. Looking forward to the PR. Btw. you'll probably want to rebase off of master before opening the PR.

**Comment by kiudee** [comment]
I’ve rebased the commits onto master and verified that the example still works.
In this state it should be ready for merging, but I welcome independent review.

**Comment by twiecki** [comment]
I haven't read the paper so I'm not sure why they focus on sampling of correlation matrices when covariance matrices should be much more direct applicable. 

I wonder if we could include some of these steps:

``` python
    corr_triangle = LKJCorrTriangle('corr', n=1, p=p)
    corr = corr_triangle[tri_index]
    corr = t.fill_diagonal(corr, 1)
    cov = Deterministic('cov', t.diag(sigma).dot(corr.dot(t.diag(sigma))))
```

Into the distribution to make it easier to use.

For example, `sigma` could be passed to `LKJ` and perhaps `LKJ` could also do the triangular filling and lastly, could just return a covariance matrix to turn all of the above into a one-liner. Thoughts?

**Comment by kiudee** [comment]
The point of the standard deviation/correlation matrix decomposition is, that one may have prior information for the standard deviations of the variables but not on their correlations.
Thus we can put a very informative prior distribution on the standard deviations while using the non-informative LKJ prior distribution on the correlations.

I agree, the current way implementation is very complicated to use. In your proposed solution the sampler will then pass a p x p matrix of which we ignore the lower half?

PS: The big problem is still that the sampler dictates what the output is, or is there a mechanism in pymc3 with which to return a different output?

**Comment by twiecki** [comment]
I see, that makes sense.

Perhaps even we could leave the LKJ prior as it now but add a new wrapper class that does the conversion. I'm not sure though.

**Comment by kiudee** [comment]
I would also prefer using this as a starting point, and then implement an elegant wrapper solution later.

**Comment by twiecki** [comment]
OK. Can you think of a way to unittest the distribution?

**Comment by kiudee** [comment]
I already looked at the existing unit tests to see how they are implemented.
Since we do not have a reference implementation, I would hard code the log probability for a few representative matrices and check if the distribution returns the correct values.
In addition, we can test the edge cases (values in support, n > 0 etc).
What would be your recommendation?

**Comment by twiecki** [comment]
Yeah, I was thinking the same.

**Comment by kiudee** [comment]
I added a few test cases for the LKJ distribution.
Is anything missing here?

**Comment by twiecki** [comment]
How did you compute the values? Would be good to add that as a comment to the test somehow.

**Comment by kiudee** [comment]
I hand calculated the values using the formulas of the paper (to avoid using the implementation which is to be tested).

**Comment by twiecki** [comment]
Great, I'll merge this once the tests pass.

**Comment by kiudee** [comment]
Hm, the build failed because of `check_mvnormal(3)` and it took `test_betabin()` a very long time to complete. Nothing related to the new code.

**Comment by twiecki** [comment]
This is a great addition to `pymc`, thanks!
